assets/js/modules/analytics-4/utils/withCustomDimensions.js

```diff
@@ -197,6 +198,40 @@ export default function withCustomDimensions( options = {} ) {
 				isNavigatingToOAuthURL ||
 				hasCustomDimensions === undefined;
 
+			const isGatheringData = useSelect( ( select ) => {
+				const isGA4GatheringData =
+					select( MODULES_ANALYTICS_4 ).isGatheringData();
+
+				if ( isGA4GatheringData !== false ) {
+					return isGA4GatheringData;
+				}
+
+				if ( loading || ! hasCustomDimensions ) {
+					// Custom dimension gathering data is not applicable if we're still loading or there are no custom dimensions.
+					return null;
+				}
+
+				const { isCustomDimensionGatheringData } =
+					select( MODULES_ANALYTICS_4 );
+
+				// Return undefined if any of the custom dimensions' gathering data state is undefined,
+				// or true if any of the custom dimensions' gathering data state is true.
+				for ( const gatheringDataState of [ undefined, true ] ) {
+					if (
+						customDimensions?.some(
+							( dimension ) =>
+								isCustomDimensionGatheringData( dimension ) ===
+								gatheringDataState
+						)
+					) {
+						return gatheringDataState;
+					}
+				}
```

Comment:
Maybe we should better add a new selector that returns the gathering state for multiple dimensions? Something like `isCustomDimensionsGatheringData`?

Reply:
Sounds good. I've created the selector as `areCustomDimensionsGatheringData()`. It basically replicates this logic, but the net result is cleaner.